### PR TITLE
Variadic parameters support for DI container

### PR DESCRIPTION
**Pull request by SamMousa**
Just tests that do not pass.

| Q             | A
| ------------- | ---
| Is bugfix?    | yes/no
| New feature?  | yes/no
| Breaks BC?    | yes/no
| Tests pass?   | no
| Fixed issues  | #14126

### CODE REVIEW

**Comment by SamMousa** [comment]
How do I mark a test for PHP7 only?

**Comment by samdark** [comment]
> How do I mark a test for PHP5.6+ only?

As far as I know, you can not except checking for PHP version and skipping manually. There are some tests like that already.

File: tests/framework/di/ContainerTest.php [hunk 1]
```diff
@@ -286,4 +287,26 @@ public function testContainerSingletons()
         $sameFoo = $container->get('qux.using.closure');
         $this->assertSame($foo, $sameFoo);
     }
+
+    /**
+     * @requires PHP 5.6
```
Comment (samdark) [inline tests/framework/di/ContainerTest.php]: Does that work?

**Review by dynasource — APPROVED** [review]

File: tests/framework/di/ContainerTest.php (hunk 1 above)
Comment (Faryshta) [inline tests/framework/di/ContainerTest.php]: https://phpunit.de/manual/current/en/incomplete-and-skipped-tests.html#incomplete-and-skipped-tests.requires.tables.api

**Review by samdark — APPROVED** [review]
Need a line for changelog.

File: tests/framework/di/ContainerTest.php (hunk 1 above)
Comment (Faryshta) [inline tests/framework/di/ContainerTest.php]: also https://github.com/sebastianbergmann/phpunit/issues/1984 this is equivalent to ask php version 5.6 or above

File: framework/di/Container.php [hunk 2]
```diff
@@ -426,7 +426,9 @@ protected function getDependencies($class)
         $constructor = $reflection->getConstructor();
         if ($constructor !== null) {
             foreach ($constructor->getParameters() as $param) {
-                if ($param->isDefaultValueAvailable()) {
+                if ($param->isVariadic()) {
```
Comment (SilverFire) [inline framework/di/Container.php]: Method `isVariadic()` is available since PHP 5.6. This change will lead to fatal error on PHP 5.4 and 5.5

Comment (SilverFire) [inline framework/di/Container.php]: This PR should go to 2.1 then

**Review by SilverFire — CHANGES_REQUESTED** [review]

File: framework/di/Container.php (hunk 2 above)
Comment (Faryshta) [inline framework/di/Container.php]: `if(method_exists($param, 'isVaridic') && $param->isVariadic())`

works on php 5.4 and 5.5 we can stll have this for 2.0.12

File: framework/di/Container.php [hunk 3]
```diff
@@ -528,7 +530,10 @@ public function resolveCallableDependencies(callable $callback, $params = [])
             $name = $param->getName();
             if (($class = $param->getClass()) !== null) {
                 $className = $class->getName();
-                if ($associative && isset($params[$name]) && $params[$name] instanceof $className) {
+                if ($param->isVariadic()) {
```
Comment (Faryshta) [inline framework/di/Container.php]: same here, add `method_exists` call

File: framework/di/Container.php (hunk 2 above)
Comment (SilverFire) [inline framework/di/Container.php]: Then I'd prefer PHP_VERSION_ID constant comparison for performance reasons

**Comment by SamMousa** [comment]
Thanks for the feedback, hope to have time for this soon..

**Review by tom-- — APPROVED** [review]
ok with the changes already requested

File: tests/framework/di/ContainerTest.php [hunk 4]
```diff
@@ -286,4 +287,26 @@ public function testContainerSingletons()
...
+        $container->get(Variadic::class);
+    }
+
+    /**
+     * @requires PHP 5.6
+     */
+    public function testVariadicCallable()
+    {
+        $container = new Container();
+        $func = function(QuxInterface ...$quxes) {
```
Comment (samdark) [inline tests/framework/di/ContainerTest.php]: Whole tests fatal-errors because of it on PHP 5.4 and PHP 5.5. Any idea on how to test it safely?

**Review by samdark — CHANGES_REQUESTED** [review]

File: tests/framework/di/ContainerTest.php (hunk 4 above)
Comment (cebe) [reply tests/framework/di/ContainerTest.php]: include the file only when PHP version is big enough. We did that for a yield test already: https://github.com/yiisoft/yii2/blob/0a7dbe3df5f40bb0e050ec01804645bfcccd4705/tests/framework/db/CommandTest.php#L307-L314

**Comment by samdark** [comment]
@SamMousa do you have time to implement [that last bit](https://github.com/yiisoft/yii2/pull/14190#discussion_r140972405) that's required in order for tests to pass?

**Comment by SamMousa** [comment]
I'm travelling so will be 2 weeks+ before i get to it, sorry!

**Comment by samdark** [comment]
Have a good trip. We'll handle it.

**Review by SilverFire — APPROVED** [review]
+ changelog

**Comment by samdark** [comment]
Merged. Thank you!

**Pull request merged**